js/tc39/tc39_test.go

```diff
@@ -0,0 +1,562 @@
+		// don't take forever
+		ctx.runTC39Test(t, name, src, meta, false)
+	}
+
+	if ctx.enableBench {
+		ctx.benchLock.Lock()
+		ctx.benchmark = append(ctx.benchmark, tc39BenchmarkItem{
+			name:     name,
+			duration: time.Since(startTime),
+		})
+		ctx.benchLock.Unlock()
+	}
+}
+
+func (ctx *tc39TestCtx) init() {
+	ctx.prgCache = make(map[string]*goja.Program)
+	ctx.errors = make(map[string]string)
+
+	b, err := ioutil.ReadFile("./breaking_test_errors.json")
+	if err != nil {
+		panic(err)
+	}
+	ctx.expectedErrors = make(map[string]string, 1000)
+	err = json.Unmarshal(b, &ctx.expectedErrors)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (ctx *tc39TestCtx) compile(base, name string) (*goja.Program, error) {
+	ctx.prgCacheLock.Lock()
+	defer ctx.prgCacheLock.Unlock()
+
+	prg := ctx.prgCache[name]
+	if prg == nil {
+		fname := path.Join(base, name)
+		f, err := os.Open(fname) //nolint:gosec
+		if err != nil {
+			return nil, err
+		}
+		defer f.Close() //nolint:gosec,errcheck
+
+		b, err := ioutil.ReadAll(f)
+		if err != nil {
+			return nil, err
+		}
+
+		str := string(b)
+		prg, _, err = ctx.compiler.Compile(str, name, "", "", false, lib.CompatibilityModeExtended)
+		if err != nil {
+			return nil, err
+		}
+		ctx.prgCache[name] = prg
+	}
+
+	return prg, nil
+}
+
+func (ctx *tc39TestCtx) runFile(base, name string, vm *goja.Runtime) error {
+	prg, err := ctx.compile(base, name)
+	if err != nil {
+		return err
+	}
+	_, err = vm.RunProgram(prg)
+	return err
+}
+
+func (ctx *tc39TestCtx) runTC39Script(name, src string, includes []string, vm *goja.Runtime) (early bool, err error) {
+	early = true
+	err = ctx.runFile(ctx.base, path.Join("harness", "assert.js"), vm)
+	if err != nil {
+		return
+	}
+
+	err = ctx.runFile(ctx.base, path.Join("harness", "sta.js"), vm)
+	if err != nil {
+		return
+	}
+
+	for _, include := range includes {
+		err = ctx.runFile(ctx.base, path.Join("harness", include), vm)
+		if err != nil {
+			return
+		}
+	}
+
+	var p *goja.Program
+	p, _, err = ctx.compiler.Compile(src, name, "", "", false, lib.CompatibilityModeExtended)
+
+	if err != nil {
+		return
+	}
+
+	early = false
+	_, err = vm.RunProgram(p)
+
+	return
+}
+
+func (ctx *tc39TestCtx) runTC39Tests(name string) {
+	files, err := ioutil.ReadDir(path.Join(ctx.base, name))
+	if err != nil {
+		ctx.t.Fatal(err)
+	}
+
+	for _, file := range files {
+		if file.Name()[0] == '.' {
+			continue
+		}
+		newName := path.Join(name, file.Name())
+		if pathBasedBlock[newName] {
+			ctx.t.Run(newName, func(t *testing.T) {
+				t.Skipf("Skip %s beause of path based block", newName)
+			})
+			continue
+		}
+		if file.IsDir() {
+			ctx.runTC39Tests(newName)
+		} else if strings.HasSuffix(file.Name(), ".js") && !strings.HasSuffix(file.Name(), "_FIXTURE.js") {
+			ctx.runTest(newName, func(t *testing.T) {
+				ctx.runTC39File(newName, t)
+			})
+		}
+	}
+}
+
+func TestTC39(t *testing.T) {
+	if testing.Short() {
```

Comment:
I had no idea this existed... And I'm not sure I even like it, given the existence of build flags. but oh well... :man_shrugging: On the other hand, I :heart: the skip message below, and how we can optionally include the tc39 tests in our normal golang tests, or not!